File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/colorselector/AlphaView.java
package net.kdt.pojavlaunch.colorselector;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Shader;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.core.math.MathUtils;

import net.kdt.pojavlaunch.Tools;

import top.defaults.checkerboarddrawable.CheckerboardDrawable;

/**
 * アルファ（透明度）値を選択するためのビュー。チェッカーボードの上に透明度グラデーションを表示します。
 */
public class AlphaView extends View {
    private final Drawable mCheckerboardDrawable = CheckerboardDrawable.create();
    private final Paint mShaderPaint = new Paint();
    private final Paint mBlackPaint;
    private final RectF mViewSize = new RectF(0,0,0,0);
    private AlphaSelectionListener mAlphaSelectionListener;
    private int mSelectedAlpha;
    private float mAlphaDiv; // 高速な位置→アルファ値変換用
    private float mScreenDiv; // 高速なアルファ値→位置変換用
    private float mWidthThird; // カーソル表示用のビュー幅の1/3

    /**
     * コンストラクタ。ペイントオブジェクトを初期化します。
     */
    public AlphaView(Context ctx, AttributeSet attrs) {
        super(ctx,attrs);
        mBlackPaint = new Paint();
        mBlackPaint.setStrokeWidth(Tools.dpToPx(3));
        mBlackPaint.setColor(Color.BLACK);
    }

    /**
     * アルファ選択リスナーを設定します。
     */
    public void setAlphaSelectionListener(AlphaSelectionListener alphaSelectionListener) {
        mAlphaSelectionListener = alphaSelectionListener;
    }

    /**
     * アルファ値を設定し、再描画します。
     */
    public void setAlpha(int alpha) {
        mSelectedAlpha = alpha;
        invalidate();
    }

    /**
     * タッチイベントを処理し、アルファ値を計算します。
     */
    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        mSelectedAlpha = (int) MathUtils.clamp(mAlphaDiv * event.getY(), 0, 0xff);
        if(mAlphaSelectionListener != null) mAlphaSelectionListener.onAlphaSelected(mSelectedAlpha);
        invalidate();
        return true;
    }

    /**
     * サイズ変更時にグラデーションと変換値を再計算します。
     */
    @Override
    protected void onSizeChanged(int w, int h, int old_w, int old_h) {
        mViewSize.right = w;
        mViewSize.bottom = h;
        mShaderPaint.setShader(new LinearGradient(0,0,0,h, 0, Color.WHITE, Shader.TileMode.REPEAT));
        mAlphaDiv = 255f / mViewSize.bottom;
        mScreenDiv = mViewSize.bottom / 255f;
        mWidthThird = mViewSize.right / 3f;
    }

    /**
     * ビューを描画します。チェッカーボード、アルファグラデーション、選択インジケーターを描画します。
     */
    @Override
    protected void onDraw(Canvas canvas) {
        mCheckerboardDrawable.draw(canvas);
        canvas.drawRect(mViewSize, mShaderPaint);
        float linePos = mSelectedAlpha * mScreenDiv;
        canvas.drawLine(0, linePos , mWidthThird, linePos, mBlackPaint);
        canvas.drawLine(mWidthThird * 2, linePos, getRight(),linePos, mBlackPaint);
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/view/colorselector/AlphaView.kt
package com.arata.yukarilauncher.ui.view.colorselector

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Shader
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import androidx.core.math.MathUtils
import com.arata.yukarilauncher.Tools
import top.defaults.checkerboarddrawable.CheckerboardDrawable

/** アルファ値（透明度）を選択するための縦方向スライダービュー。 */
class AlphaView @JvmOverloads constructor(ctx: Context, attrs: AttributeSet? = null) : View(ctx, attrs) {
    /** 透明部分を表現するチェッカーボード描画用Drawable。 */
    private val mCheckerboardDrawable: Drawable = CheckerboardDrawable.create()
    /** シェーダー（白→透明のグラデーション）描画用ペイント。 */
    private val mShaderPaint = Paint()
    /** 選択位置を示す線のペイント。 */
    private val mBlackPaint: Paint
    /** ビューの矩形領域。 */
    private val mViewSize = RectF(0f, 0f, 0f, 0f)
    /** アルファ選択リスナー。 */
    private var mAlphaSelectionListener: AlphaSelectionListener? = null
    /** 現在選択中のアルファ値。 */
    private var mSelectedAlpha = 0
    /** ピクセル→アルファ値 変換係数。 */
    private var mAlphaDiv = 0f
    /** アルファ値→ピクセル 変換係数。 */
    private var mScreenDiv = 0f
    /** ビュー幅の3分の1。 */
    private var mWidthThird = 0f

    init {
        mBlackPaint = Paint().apply {
            strokeWidth = Tools.dpToPx(3f)
            color = Color.BLACK
        }
    }

    /** アルファ選択リスナーを設定する。 */
    fun setAlphaSelectionListener(alphaSelectionListener: AlphaSelectionListener?) {
        mAlphaSelectionListener = alphaSelectionListener
    }

    /** アルファ値を設定し、ビューを再描画する。 @param alpha 設定するアルファ値（0～255） */
    fun setAlpha(alpha: Int) {
        mSelectedAlpha = alpha
        invalidate()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        mSelectedAlpha = MathUtils.clamp(mAlphaDiv * event.y, 0.0f, 255.0f).toInt()
        mAlphaSelectionListener?.onAlphaSelected(mSelectedAlpha)
        invalidate()
        return true
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        mViewSize.right = w.toFloat()
        mViewSize.bottom = h.toFloat()
        mShaderPaint.shader = LinearGradient(0f, 0f, 0f, h.toFloat(), 0, Color.WHITE, Shader.TileMode.REPEAT)
        mAlphaDiv = 255f / mViewSize.bottom
        mScreenDiv = mViewSize.bottom / 255f
        mWidthThird = mViewSize.right / 3f
    }

    override fun onDraw(canvas: Canvas) {
        mCheckerboardDrawable.draw(canvas)
        canvas.drawRect(mViewSize, mShaderPaint)
        val linePos = mSelectedAlpha * mScreenDiv
        canvas.drawLine(0f, linePos, mWidthThird, linePos, mBlackPaint)
        canvas.drawLine(mWidthThird * 2, linePos, right.toFloat(), linePos, mBlackPaint)
    }
}
